server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1204,22 +1206,29 @@ private void cleanupStaleBlobs(
             listener.onResponse(deleteResult);
         }, listener::onFailure));
 
-        final Executor executor = threadPool.executor(ThreadPool.Names.SNAPSHOT);
         final List<String> staleRootBlobs = staleRootBlobs(newRepoData, rootBlobs.keySet());
         if (staleRootBlobs.isEmpty()) {
             groupedListener.onResponse(DeleteResult.ZERO);
         } else {
-            executor.execute(ActionRunnable.supply(groupedListener, () -> {
```

Comment:
nit/suggestion (sorry Github won't let use its suggestion feature here):

```diff
diff --git a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
index dfdc09b696f..83bcf70fe2f 100644
--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
@@ -1210,25 +1210,26 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
         if (staleRootBlobs.isEmpty()) {
             groupedListener.onResponse(DeleteResult.ZERO);
         } else {
-            snapshotDeleteTaskRunner.enqueueTask(ActionListener.wrap(releasable -> {
+            snapshotDeleteTaskRunner.enqueueTask(groupedListener.map(releasable -> {
                 try (releasable) {
                     List<String> deletedBlobs = cleanupStaleRootFiles(newRepoData.getGenId() - 1, deletedSnapshots, staleRootBlobs);
-                    groupedListener.onResponse(
-                        new DeleteResult(deletedBlobs.size(), deletedBlobs.stream().mapToLong(name -> rootBlobs.get(name).length()).sum())
+                    return new DeleteResult(
+                        deletedBlobs.size(),
+                        deletedBlobs.stream().mapToLong(name -> rootBlobs.get(name).length()).sum()
                     );
                 }
-            }, groupedListener::onFailure));
+            }));
         }

         final Set<String> survivingIndexIds = newRepoData.getIndices().values().stream().map(IndexId::getId).collect(Collectors.toSet());
         if (foundIndices.keySet().equals(survivingIndexIds)) {
             groupedListener.onResponse(DeleteResult.ZERO);
         } else {
-            snapshotDeleteTaskRunner.enqueueTask(ActionListener.wrap(releasable -> {
+            snapshotDeleteTaskRunner.enqueueTask(groupedListener.map(releasable -> {
                 try (releasable) {
-                    groupedListener.onResponse(cleanupStaleIndices(foundIndices, survivingIndexIds));
+                    return cleanupStaleIndices(foundIndices, survivingIndexIds);
                 }
-            }, groupedListener::onFailure));
+            }));
         }
     }
```


NB slightly different semantics, this means we release `releasable` (potentially starting another delete task) before completing the grouped listener. I think that's ok tho.